xwayland/sockets.c

```diff
@@ -84,11 +84,44 @@ static int open_socket(struct sockaddr_un *addr, size_t path_size) {
 	return -1;
 }
 
+static bool check_socket_dir(void) {
+	struct stat buf;
+
+	if (lstat(socket_dir, &buf)) {
+		wlr_log_errno(WLR_ERROR, "Failed to stat %s", socket_dir);
+		return false;
+	}
+	if (!(buf.st_mode & S_IFDIR)) {
+		wlr_log(WLR_ERROR, "%s is not a directory", socket_dir);
+		return false;
+	}
+	if (!((buf.st_uid == 0) || (buf.st_uid == getuid()))) {
+		wlr_log(WLR_ERROR, "%s not owned by root or us", socket_dir);
+		return false;
+	}
+	if (!(buf.st_mode & S_ISVTX)) {
+		if ((buf.st_mode & (S_IWGRP | S_IWOTH))) {
+			wlr_log(WLR_ERROR, "sticky bit not set on %s", socket_dir);
+			return false;
+		}
+	}
+	return true;
+}
+
 static bool open_sockets(int socks[2], int display) {
 	struct sockaddr_un addr = { .sun_family = AF_UNIX };
 	size_t path_size;
 
-	mkdir(socket_dir, 0777);
+	if (mkdir(socket_dir, 0755) == 0) {
```

Comment:
Is it dangerous to not set the sticky bit here?

Mutter seems to create the socket with mode 01777, and with a non-root owner. This seems dangerous, since this will potentially allow the current user to mess up other users' X11 sockets.

Setting the mode 0755 makes it clear that the directory is to be used exclusively by the current user, since other users won't be able to create files in it.

So I think our code is safer. Am I missing something?

Reply:
If the current user is the owner, they are never going to be hindered by any mode as they can change it at will. The advantage here is rather in avoiding the trouble of setting the sticky bit portably without ignoring a more restrictive umask (which I always tend to find rather distasteful) if we're content catering to the simplest single user setups (and will ourselves refuse to use another non-root socket directory regardless).